### PR TITLE
Feature #2168 tc_diag

### LINKED ISSUE #2168: Create an initial development version of a new TC-Diag tool to support the computation of tropical cyclone diagnostics
## Describe the New Feature ##
The METplus TC-Diagnostics group met on May 23, 2022 to define requirements for incorporating legacy hurricane diagnostics logic in METplus. Those requirement include the creation of a new tool named TC-Diag, for Tropical Cyclone Diagnostics.

The inputs for TC-Diag are the same as those for TC-RMW:
  - a list of gridded input data files (-data)
  - an ATCF track data file (-deck)
  - a configuration file (-config)
  - an output file name (-out)

For each track point, TC-Diag transforms the gridded model data to cylindrical coordinates centered on the storm location. This can be done relative to the radius of maximum winds-space, as is done in TC-RMW. However, additional conversion logic should also be supported, such as being based on shear.

ACTION: Need to define what derivation types should be supported.

After the requested fields are read and transformed in one or more ways, provide an option to write them to an output NetCDF file. However, recommend that this option be disabled by default.

Next, incorporate python code from Robert DeMaria directly into the MET repository to actually compute the diagnostics. The TC-Diag makes calls to that python code to derive the diagnostics. The number and type of diagnostics to be computed may depend on the type of cylindrical coordinate transformation that was performed. Ideally the cylindrical coordinate data is handed to python in memory. This logic should be made highly configurable. Provide a mechanism for users to write or modify a python derivation script.

ACTION: Need to define lots of details for how MET calls the python derivation scripts and how this can be made configurable. Also need to define source of those python scripts.

The python scripts hand back to TC-Diag the computed diagnostics. These will consist of a few dozen floating point numbers for each track location.

TC-Diag writes those diagnostic values out to one of two supported formats. The desired output format must be a configurable option:

1. LS Diag (Large Scale Diagnostics) format: See sample [LS Diag formatted](https://rammb2.cira.colostate.edu/research/tropical-cyclones/ships/) files only the [NHC FTP site](https://ftp.nhc.noaa.gov/atcf/lsdiag/).
2. Model Diag format: Need to define documentation and sample data.

### Acceptance Testing ###
Input: gridded model data, ATCF track file, ASCII config file
Output: Formatted ASCII output in the LS-Diag or Model-Diag formats

### Time Estimate ###
2 weeks.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
The initial version of the TC-Diag tool should be ready by February 2023.

### Funding Source ###
*NOAA WPO - JTTI TCDiag Project (2770043)*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: @JohnHalleyGotway 
- [x] Select **scientist(s)** or **no scientist** required: @jvigh and @KathrynNewman 

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] MET: dtcenter/MET#392
- [ ] METplus: List issue here
- [ ] METdatadb: List issue here
- [ ] METviewer: List issue here

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
This introduces the first version of the TC-Diag tool. This initial version is limited in functionality and only actually writes gridded NetCDF output in cylindrical coordinates. It does NOT actually compute TC diagnostics yet, but will do so in MET version 12.0.0.

## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
Adds new TC-Diag tool.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>
Adds new NetCDF output from TC-Diag in cylindrical coordinates. It is pretty similar to the output from TC-RMW.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Tested extensively during development.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review code changes.
Review doc updates - although the actual doc contents will be updated by #2558.
Inspect the new output from the new unit_tc_diag.xml test.
Run on the command line and note the warning message about the "beta status".

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

While this does introduce a new TC-Diag chapter, I will review/revise it with #2558.

- [x] Do these changes include sufficient testing updates? **[Yes]**

Yes, for now. We will want to add new tests and/or modify the existing one after the TC-Diag tool is more complete.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

Adds new output from `unit_tc_diag.xml`.

- [x] Please complete this pull request review by **[Friday 6/9/23]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** issue with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: docs/Users_Guide/tc-diag.rst [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+The default configuration file for the TC-Diag tool named **TCDiagConfig_default** can be found in the installed *share/met/config/* directory. It is encouraged for users to copy these default files before modifying their contents. The contents of the configuration file are described in the subsections below.
+
+The TC-Diag tool should be configured to filter the input track data (**-deck**) down to the subset of tracks that correspond to the gridded data files provided (**-data**). The filtered tracks should contain data for only one initialization time but may contain tracks for multiple models.
+
+_______________________
+
+.. code-block:: none
+
+  model = [ "GFSO", "OFCL" ];
```
Comment (jvigh) [inline docs/Users_Guide/tc-diag.rst:61]: It seems that the adecks from NHC currently use AVNO, not GFS0. Should this default config value be changed to match the current practice in use at NHC?

Comment (JohnHalleyGotway) [reply docs/Users_Guide/tc-diag.rst:61]: When you run TC-Diag with this configuration, you'll see the following log message:
```
DEBUG 1: When reading ATCF track data, all instances of "AVN" are automatically replaced with "GFS" (ATCF ID = AVNO).
```
All instances of `AVN` are auto-replaced by `GFS`. But this is existing logic unrelated to the TC-Diag changes. Its fine to discuss this logic but I'd recommend separating it from this TC-Diag pull request.

Also, here's the warning message that's printed every time its run:
```
WARNING: 
WARNING: The TC-Diag tool is provided in BETA status for MET V11.1.0.
WARNING: Please see the release notes of future MET versions for updates.
WARNING: 
```

File: docs/Users_Guide/tc-diag.rst [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+  rmw_scale = 0.2;
+
+The **rmw_scale** parameter overrides the **max_range_km** parameter. When this is set the radial grid spacing will be **rmw_scale** in units of the RMW, which varies along the storm track.
+
+_______________________
+
+.. code-block:: none
+
+  compute_tangential_and_radial_winds = TRUE;
```
Comment (jvigh) [inline docs/Users_Guide/tc-diag.rst:145]: Just for clarity, is compute_tangential_and_radial_winds = False; ever used in the TCDiag tool? Do any of the current diagnostics use U/V winds?

File: docs/Users_Guide/tc-diag.rst (hunk 1 above)
Comment (jvigh) [reply docs/Users_Guide/tc-diag.rst:61]: Sure. Maybe that should be a separate issue. I just don't why we're still using GFSO as the default config value when that has long ago fallen out of practice.

File: src/basic/vx_config/config_constants.h [hunk 3]
```diff
@@ -1193,6 +1193,18 @@ static const char conf_key_radial_velocity_field_name[] = "radial_velocity_field
 static const char conf_key_tangential_velocity_field_name[] = "tangential_velocity_field_name";
 static const char conf_key_radial_velocity_long_field_name[] = "radial_velocity_long_field_name";
 static const char conf_key_tangential_velocity_long_field_name[] = "tangential_velocity_long_field_name";
+static const char conf_key_vortex_removal[] = "vortex_removal";
+
+//
+// TC-Diag specific paramter key names
```
Comment (jvigh) [inline src/basic/vx_config/config_constants.h]: Minor mispelling here: paramter -> parameter

File: src/tools/tc_utils/tc_diag/tc_diag_conf_info.cc [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+   clear();
+
+   // Note: tech_ids and data_files are specified on the
+   //       command line rather than in the config file
+
+   // Conf: domain
+   domain = dict.lookup_string(conf_key_domain);
+
+   // Hard-code the name
+   data.name = "TCRMW";
```
Comment (jvigh) [inline src/tools/tc_utils/tc_diag/tc_diag_conf_info.cc]: Should this be TCDIAG? Or is the data still considered "TCRMW"?

**Review by jvigh — APPROVED** [review]
@JohnHalleyGotway I have finished reviewing all of the code changes. They look good to me. I just left a few clarification questions and spotted a misspelling in one spot. This development represents a ton of work. Good job at getting this part of the finish line!

**Review by jvigh — APPROVED** [review]
@JohnHalleyGotway  I have left some questions/comments for clarification. I have finished checking all of the code changes. This looks very good. It represents a ton of work, so congratulations on getting this first phase of the TCDiag tool across the finish line. !

**Comment by jvigh** [comment]
Sorry for the duplicate messages. I thought perhaps I had not done my
review correctly, but now I just see that it adds it to the conversation.

File: docs/Users_Guide/tc-diag.rst (hunk 2 above)
Comment (JohnHalleyGotway) [reply docs/Users_Guide/tc-diag.rst:145]: So this does not actually appear in the TC-Diag config file anymore. I'll clean up the docs after the RC1 release for #2558.

File: src/basic/vx_config/config_constants.h (hunk 3 above)
Comment (JohnHalleyGotway) [reply src/basic/vx_config/config_constants.h]: Good catch. Its now fixed with this [commit](https://github.com/dtcenter/MET/commit/b6ee12a6c111928772c055105f96b823d1341a20).

File: src/tools/tc_utils/tc_diag/tc_diag_conf_info.cc (hunk 4 above)
Comment (JohnHalleyGotway) [reply src/tools/tc_utils/tc_diag/tc_diag_conf_info.cc]: I originally thought that hard-coded string was necessary, but I switched it to `TCDIAG` with this [commit](https://github.com/dtcenter/MET/commit/b6ee12a6c111928772c055105f96b823d1341a20).

File: docs/Users_Guide/tc-diag.rst (hunk 1 above)
Comment (JohnHalleyGotway) [reply docs/Users_Guide/tc-diag.rst:61]: Let's talk more about this at the next TC-Diag meeting and @KathrynNewman can weigh in.

**Comment by JohnHalleyGotway** [comment]
Thanks for the review @jvigh. Proceeding with merging these changes.

**Pull request merged**